File: H08/handson8_2.m
% %Estimar SNR de sinais recebidos com ruido
% %Tenho informações do sinal enviado (Am,fm) -> consigo encontrar a potencia
% %do sinal enviado
% %Calculo a potencia do sinal recebido (y=x+n)
% % n=y-x ...
% 
% clear all; clc; close all;
% load('Pratica_08_sinal_real.mat')
% N=length(y);
% t=0:1/fs:(N-1)/fs;   
% x_orig=Am*cos(2*pi*fm*t); %Sinal original
% %Calcular Potência do Sinal Original:
% Px = sum(abs(x_orig).^2)/N;
% %O sinal y corresponde a y=x+n -> n=y-x
% n=y-x_orig;
% %Potencia do ruido n:
% Pn=sum(abs(n).^2)/N;
% %Determinando SNR:
% SNR=Px/Pn;
% SNR_dB=10*log10(SNR);
% %%
% %Plotagens
% figure(1)
% subplot(3,1,1);
% plot(t,x_orig);
% title('Sinal Original');
% xlabel('t[s]');
% subplot(3,1,2);
% plot(t,y);
% title(['Sinal Recebido - SNR = ' num2str(SNR_dB) 'dB']);
% xlabel('t[s]');
% ylim([-2 2])
% subplot(3,1,3);
% plot(t,n);
% xlabel('t[s]');
% title('Ruído');
% 
%Estimar SNR de sinais recebidos com ruido
%Tenho informações do sinal enviado (Am,fm) -> consigo encontrar a potencia
%do sinal enviado
%Calculo a potencia do sinal recebido (y=x+n)
% n=y-x ...

clear all; clc; close all;
load('Pratica_08_sinal_complexo.mat')
N=length(y);
t=0:1/fs:(N-1)/fs;   
x_orig=Ar*cos(2*pi*fm*t)+1i*Ai*cos(2*pi*fm*t); %Sinal original
%Calcular Potência do Sinal Olriginal:
Px = sum(abs(x_orig).^2)/N;
%O sinal y corresponde a y=x+n -> n=y-x
n=y-x_orig;
%Potencia do ruido n:
Pn=sum(abs(n).^2)/N;
%Determinando SNR:
SNR=Px/Pn;
SNR_dB=10*log10(SNR);
%%
%Plotagens
figure(1)
subplot(3,1,1);
plot(t,abs(x_orig));
title('Sinal Original');
xlabel('t[s]');
subplot(3,1,2);
plot(t,abs(y));
title(['Sinal Recebido - SNR = ' num2str(SNR_dB) 'dB']);
xlabel('t[s]');
%ylim([-2 2])
subplot(3,1,3);
plot(t,n);
xlabel('t[s]');
title('Ruído');

File: H08/handson8_2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon May 27 10:39:41 2019

@author: fellipe
"""

#from scipy.io import loadmat
#import numpy as np
#import matplotlib.pyplot as plt
##Carregando parametros
#variaveis = loadmat('./Pratica_08_sinal_real.mat')
#Am=int(variaveis['Am'])
#fm=int(variaveis['fm'])
#fs=int(variaveis['fs'])
#y=variaveis['y'].flatten()
#N=len(y)
#t=np.arange(0,N/fs,1/fs)
##Construindo Sinal Original
#x_orig=Am*np.cos(2*np.pi*fm*t)
##Calculando Potencia do sinal original
#P_x=(np.linalg.norm(x_orig)**2)/N
##Como y=x+n => n=y-x
#n=y-x_orig
#P_n=(np.linalg.norm(n)**2)/N
#SNR=P_x/P_n
#SNR_dB=10*np.log10(SNR)
##Plotagens
#plt.figure(1,[20,20])
#plt.subplot(3,1,1)
#plt.plot(t,x_orig)
#plt.title('Sinal Original')
#plt.xlabel('t[s]')
#plt.subplot(3,1,2)
#plt.plot(t,y)
#plt.title('Sinal Recebido: SNR = ' + str(SNR_dB) + 'dB')
#plt.xlabel('t[s]')
#plt.subplot(3,1,3)
#plt.plot(t,n)
#plt.title('Ruído AWGN')
#plt.xlabel('t[s]')
#plt.show()
from scipy.io import loadmat
import numpy as np
import matplotlib.pyplot as plt
#Carregando parametros
variaveis = loadmat('./Pratica_08_sinal_complexo.mat')
Ar=int(variaveis['Ar'])
Ai=int(variaveis['Ai'])
fm=int(variaveis['fm'])
fs=int(variaveis['fs'])
y=variaveis['y'].flatten()
N=len(y)
t=np.arange(0,N/fs,1/fs)
#Construindo Sinal Original
x_orig=Ar*np.cos(2*np.pi*fm*t)+1j*Ai*np.cos(2*np.pi*fm*t)
#Calculando Potencia do sinal original
P_x=(np.linalg.norm(x_orig)**2)/N
#Como y=x+n => n=y-x
n=y-x_orig
P_n=(np.linalg.norm(n)**2)/N
SNR=P_x/P_n
SNR_dB=10*np.log10(SNR)
#Plotagens
plt.figure(1,[20,20])
plt.subplot(3,1,1)
plt.plot(t,np.abs(x_orig))
plt.title('Sinal Original')
plt.xlabel('t[s]')
plt.subplot(3,1,2)
plt.plot(t,np.abs(y))
plt.title('Sinal Recebido: SNR = ' + str(SNR_dB) + 'dB')
plt.xlabel('t[s]')
plt.subplot(3,1,3)
plt.plot(t,np.abs(n))
plt.title('Ruído AWGN')
plt.xlabel('t[s]')
plt.show()




